storage/remote/marker_handler.go

```diff
@@ -0,0 +1,115 @@
+package remote
+
+import (
+	"github.com/prometheus/prometheus/tsdb/wlog"
+)
+
+type MarkerHandler interface {
+	wlog.Marker
+
+	UpdateReceivedData(segmentId, dataCount int) // Data queued for sending
+	UpdateSentData(segmentId, dataCount int)     // Data which was sent or given up on sending
+
+	Stop()
+}
+
+type markerHandler struct {
+	markerFileHandler MarkerFileHandler
+	lastMarkedSegment int
+	dataIOUpdate      chan data
+	quit              chan struct{}
+}
+
+// TODO: Rename this struct
+type data struct {
+	segmentId int
+	dataCount int
+}
+
+var (
+	_ MarkerHandler = (*markerHandler)(nil)
+)
+
+func NewMarkerHandler(mfh MarkerFileHandler) MarkerHandler {
+	mh := &markerHandler{
+		lastMarkedSegment: -1, // Segment ID last marked on disk.
+		markerFileHandler: mfh,
+		//TODO: What is a good size for the channel?
+		dataIOUpdate: make(chan data, 100),
+		quit:         make(chan struct{}),
+	}
+
+	// Load the last marked segment from disk (if it exists).
+	if lastSegment := mh.markerFileHandler.LastMarkedSegment(); lastSegment >= 0 {
+		mh.lastMarkedSegment = lastSegment
+	}
+
+	//TODO: Should this be in a separate Start() function?
+	go mh.updatePendingData()
+
+	return mh
+}
+
+func (mh *markerHandler) LastMarkedSegment() int {
+	return mh.markerFileHandler.LastMarkedSegment()
+}
+
+func (mh *markerHandler) UpdateReceivedData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- data{
+		segmentId: segmentId,
+		dataCount: dataCount,
+	}
+}
+
+func (mh *markerHandler) UpdateSentData(segmentId, dataCount int) {
+	mh.dataIOUpdate <- data{
+		segmentId: segmentId,
+		dataCount: -1 * dataCount,
+	}
+}
+
+func (mh *markerHandler) Stop() {
+	// Firstly stop the Marker Handler, because it might want to use the Marker File Handler.
+	mh.quit <- struct{}{}
+
+	// Finally, stop the File Handler.
+	mh.markerFileHandler.Stop()
+}
+
+// updatePendingData updates a counter for how much data is yet to be sent from each segment.
+// "dataCount" will be added to the segment with ID "dataSegment".
+func (mh *markerHandler) updatePendingData() {
+	batchSegmentCount := make(map[int]int)
+
+	for {
+		select {
+		case <-mh.quit:
+			return
+		case dataUpdate := <-mh.dataIOUpdate:
+			batchSegmentCount[dataUpdate.segmentId] += dataUpdate.dataCount
+		}
+
+		markableSegment := -1
+		for segment, count := range batchSegmentCount {
+			//TODO: If count is less than 0, then log an error and remove the entry from the map?
+			if count != 0 {
+				continue
+			}
+
+			//TODO: Is it save to assume that just because a segment is 0 inside the map,
+			//      all samples from it have been processed?
```

Comment:
I think the answer to this is "no". It might be possible for the watcher to send samples and then for the remote write to send those samples before the watcher is able to send new ones. This would drop the counter to 0, even though the watcher is still on the same segment.

The implementation in [#9862](https://github.com/prometheus/prometheus/pull/9862) doesn't have this issue because it only updates the marker file when the watcher switches to a new segment, and never writes the current segment in the marker file. I could try to do something similar here? I think it'd still have a couple of issue though:

* I'm not sure how this would work with offsets, if we are to use them someday
* It makes an assumption that the watcher will always use monotonically increasing segment IDs. That's probably a very safe assumption, but not sure if it's documented

Maybe it would be best to update the `WriteTo` interface in the watcher with a new function to indicate that the watcher has finished with a certain segment?

Reply:
It depends on the mechanism for decrementing the counter in the map. If we increment when we enqueue samples from the watcher into the queue manager and only decrement after successfully sending samples, then there's a way this could work. If you decrement something to 0 in the map, and then check if samples are enqueued for a segment ID higher than that one that's 0, we can assume all samples from that segment have been processed and sent.

We could also introduce a mechanism for notifying the `WriteTo` of the segment that's being read currently.

> I'm not sure how this would work with offsets, if we are to use them someday

lets worry about this in the future if we get to the point that we want/need offsets, already we've been talking about this feature for ~2y without merging the basic iteration of it

> It makes an assumption that the watcher will always use monotonically increasing segment IDs. That's probably a very safe assumption, but not sure if it's documented

this is the case, missing segments (having segments 1-10, missing 11, and then having 12+ is considered a failure scenario for the WAL: https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/wlog.go#L917

Reply:
Popping in since I'm interested in doing something similar for the logs (Promtail) WAL. Small Q for the moment.

> I'm not sure how this would work with offsets, if we are to use them someday
With this you mean to keep track of the last offset inside a segment being read?

Reply:
> With this you mean to keep track of the last offset inside a segment being read?

Yes, exactly. In [#9862](https://github.com/prometheus/prometheus/pull/9862) it was mentioned as something we will probably want in the future.